google-auth-provider/main.go

```diff
@@ -0,0 +1,185 @@
+	})
+	mux.HandleFunc("/obot-get-state", obotGetState(oauthProxy))
+	mux.HandleFunc("/obot-get-icon-url", obotGetIconURL)
+	mux.HandleFunc("/", oauthProxy.ServeHTTP)
+
+	fmt.Printf("listening on 127.0.0.1:%s\n", port)
+	if err := http.ListenAndServe("127.0.0.1:"+port, mux); err != nil {
+		fmt.Printf("failed to listen and serve: %v\n", err)
+		os.Exit(1)
+	}
+}
+
+type SerializableRequest struct {
+	Method string              `json:"method"`
+	URL    string              `json:"url"`
+	Header map[string][]string `json:"header"`
+}
+
+type SerializableState struct {
+	ExpiresOn         *time.Time `json:"expiresOn"`
+	AccessToken       string     `json:"accessToken"`
+	PreferredUsername string     `json:"preferredUsername"`
+	User              string     `json:"user"`
+	Email             string     `json:"email"`
+}
+
+func obotGetState(p *oauth2proxy.OAuthProxy) http.HandlerFunc {
+	return func(w http.ResponseWriter, r *http.Request) {
+		var sr SerializableRequest
+		if err := json.NewDecoder(r.Body).Decode(&sr); err != nil {
+			http.Error(w, fmt.Sprintf("failed to decode request body: %v", err), http.StatusBadRequest)
+			return
+		}
+
+		reqObj, err := http.NewRequest(sr.Method, sr.URL, nil)
+		if err != nil {
+			http.Error(w, fmt.Sprintf("failed to create request object: %v", err), http.StatusBadRequest)
+			return
+		}
+
+		for k, v := range sr.Header {
+			reqObj.Header[k] = v
+		}
+
+		state, err := p.LoadCookiedSession(reqObj)
+		if err != nil {
+			http.Error(w, fmt.Sprintf("failed to load cookied session: %v", err), http.StatusBadRequest)
+			return
+		}
+
+		if state == nil {
+			http.Error(w, "state is nil", http.StatusInternalServerError)
+			return
+		}
+
+		ss := SerializableState{
+			ExpiresOn:         state.ExpiresOn,
+			AccessToken:       state.AccessToken,
+			PreferredUsername: state.PreferredUsername,
+			User:              state.User,
+			Email:             state.Email,
+		}
+
+		if err := json.NewEncoder(w).Encode(ss); err != nil {
+			http.Error(w, fmt.Sprintf("failed to encode state: %v", err), http.StatusInternalServerError)
+			return
+		}
+	}
+}
+
+func obotGetIconURL(w http.ResponseWriter, r *http.Request) {
+	auth := r.Header.Get("Authorization")
+	if auth == "" {
+		http.Error(w, "missing Authorization header", http.StatusBadRequest)
+		return
+	}
+
+	accessToken := strings.TrimPrefix(auth, "Bearer ")
+	if accessToken == "" {
+		http.Error(w, "missing access token", http.StatusBadRequest)
+		return
+	}
+
+	iconURL, err := profile.FetchGoogleProfileIconURL(r.Context(), accessToken)
+	if err != nil {
+		http.Error(w, fmt.Sprintf("failed to fetch profile icon URL: %v", err), http.StatusBadRequest)
+		return
+	}
+
+	type response struct {
+		IconURL string `json:"iconURL"`
+	}
+
+	if err := json.NewEncoder(w).Encode(response{IconURL: iconURL}); err != nil {
+		http.Error(w, fmt.Sprintf("failed to encode response: %v", err), http.StatusInternalServerError)
+		return
+	}
+
+	w.WriteHeader(http.StatusOK)
```

Comment:
Same comment about this already being written when we write `w`.

Reply:
done